internal/jujuapi/http_proxy_integration_test.go

```diff
@@ -0,0 +1,228 @@
+// Copyright 2024 Canonical.
+
+package jujuapi_test
+
+import (
+	"context"
+	"encoding/pem"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"strings"
+
+	"github.com/juju/charm/v12"
+	"github.com/juju/charm/v12/resource"
+	"github.com/juju/juju/api"
+	"github.com/juju/juju/api/client/charms"
+	"github.com/juju/juju/api/client/resources"
+	"github.com/juju/juju/core/network"
+	jujuparams "github.com/juju/juju/rpc/params"
+	"github.com/juju/juju/testcharms"
+	"github.com/juju/juju/testing/factory"
+	"github.com/juju/names/v5"
+	"github.com/juju/version/v2"
+	gc "gopkg.in/check.v1"
+
+	"github.com/canonical/jimm/v3/internal/jujuapi"
+)
+
+type httpProxySuite struct {
+	websocketSuite
+}
+
+var _ = gc.Suite(&httpProxySuite{})
+
+func (s *httpProxySuite) TestHTTPAuthenticate(c *gc.C) {
+	ctx := context.Background()
+	httpProxier := jujuapi.HTTPProxier(s.JIMM)
+
+	// good
+	req, err := http.NewRequest("POST", fmt.Sprintf("/%s/charms", s.Model.UUID.String), nil)
+	c.Assert(err, gc.IsNil)
+	token, err := s.JIMM.OAuthAuthenticator.MintSessionToken(s.AdminUser.Name)
+	c.Assert(err, gc.IsNil)
+	req.SetBasicAuth("", token)
+	c.Assert(err, gc.IsNil)
+	err = httpProxier.Authenticate(ctx, nil, req)
+	c.Assert(err, gc.IsNil)
+
+	// missing auth
+	req, err = http.NewRequest("POST", fmt.Sprintf("/%s/charms", s.Model.UUID.String), nil)
+	c.Assert(err, gc.IsNil)
+	err = httpProxier.Authenticate(ctx, nil, req)
+	c.Assert(err, gc.ErrorMatches, "authentication missing")
+
+	// wrong user
+	req, err = http.NewRequest("POST", fmt.Sprintf("/%s/charms", s.Model.UUID.String), nil)
+	c.Assert(err, gc.IsNil)
+	token, err = s.JIMM.OAuthAuthenticator.MintSessionToken("test-user")
+	c.Assert(err, gc.IsNil)
+	req.SetBasicAuth("", token)
+	c.Assert(err, gc.IsNil)
+	err = httpProxier.Authenticate(ctx, nil, req)
+	c.Assert(err, gc.ErrorMatches, "unauthorized")
+}
+
+func (s *httpProxySuite) TestCharmHTTPServe(c *gc.C) {
+	ctx := context.Background()
+	httpProxier := jujuapi.HTTPProxier(s.JIMM)
+	expectU, expectP, err := s.JIMM.GetCredentialStore().GetControllerCredentials(ctx, s.Model.Controller.Name)
+	c.Assert(err, gc.IsNil)
+	// we expect the controller to respond with TLS
+	fakeController := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		u, p, _ := r.BasicAuth()
+		c.Assert(u, gc.Equals, names.NewUserTag(expectU).String())
+		c.Assert(p, gc.Equals, expectP)
+		_, err = w.Write([]byte("OK"))
+		c.Assert(err, gc.IsNil)
+	}))
+	defer fakeController.Close()
+	controller := s.Model.Controller
+	pemData := pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: fakeController.Certificate().Raw,
+	})
+	controller.CACertificate = string(pemData)
+
+	tests := []struct {
+		description    string
+		setup          func()
+		url            string
+		statusExpected int
+	}{
+		{
+			description: "good",
+			setup: func() {
```

Comment:
Some comments on these tests,
- We're misusing the websocket suite by modifying the Juju controller that is created and replacing its details with the fake controller. This is unclear and we should avoid it.
- We're not really testing the /model/UUID/charms endpoint here so much as we are testing connections to a Juju controller. E.g. Why would we test here what happens when trying to connect to a controller if the controller is unreachable. I can see the value is testing that the http proxy connects to the Juju controller but I'm not sure about these tests. I'm inclined to think you could delete them.

Reply:
These tests feel like they're better suited for proxy_http.go